packages/ledgerstate/input.go

```diff
@@ -0,0 +1,270 @@
+func InputFromMarshalUtil(marshalUtil *marshalutil.MarshalUtil) (input Input, err error) {
+	inputType, err := marshalUtil.ReadByte()
+	if err != nil {
+		err = xerrors.Errorf("failed to parse InputType (%v): %w", err, ErrParseBytesFailed)
+		return
+	}
+	marshalUtil.ReadSeek(-1)
+
+	switch InputType(inputType) {
+	case UTXOInputType:
+		if input, err = UTXOInputFromMarshalUtil(marshalUtil); err != nil {
+			err = xerrors.Errorf("failed to parse UTXOInput from MarshalUtil: %w", err)
+			return
+		}
+	default:
+		err = xerrors.Errorf("unsupported InputType (%X): %w", inputType, ErrParseBytesFailed)
+		return
+	}
+
+	return
+}
+
+// endregion ///////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// region Inputs ///////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// Inputs represents a collection of Inputs that ensures a deterministic order.
+type Inputs []Input
+
+// NewInputs returns a deterministically ordered collection of Inputs removing existing duplicates.
+func NewInputs(optionalInputs ...Input) (inputs Inputs) {
+	seenInputs := make(map[string]types.Empty)
+	sortedInputs := make([]struct {
+		input           Input
+		inputSerialized []byte
+	}, 0)
+
+	// filter duplicates (store marshaled version so we don't need to marshal a second time during sort)
+	for _, input := range optionalInputs {
+		marshaledInput := input.Bytes()
+		marshaledInputAsString := typeutils.BytesToString(marshaledInput)
+
+		if _, seenAlready := seenInputs[marshaledInputAsString]; seenAlready {
+			continue
+		}
+		seenInputs[marshaledInputAsString] = types.Void
+
+		sortedInputs = append(sortedInputs, struct {
+			input           Input
+			inputSerialized []byte
+		}{input, marshaledInput})
+	}
+
+	// sort inputs
+	sort.Slice(sortedInputs, func(i, j int) bool {
+		return bytes.Compare(sortedInputs[i].inputSerialized, sortedInputs[j].inputSerialized) < 0
+	})
+
+	// create result
+	inputs = make(Inputs, len(sortedInputs))
+	for i, sortedInput := range sortedInputs {
+		inputs[i] = sortedInput.input
+	}
+
+	return
+}
+
+// InputsFromBytes unmarshals a collection of Inputs from a sequence of bytes.
+func InputsFromBytes(inputBytes []byte) (inputs Inputs, consumedBytes int, err error) {
+	marshalUtil := marshalutil.New(inputBytes)
+	if inputs, err = InputsFromMarshalUtil(marshalUtil); err != nil {
+		err = xerrors.Errorf("failed to parse Inputs from MarshalUtil: %w", err)
+		return
+	}
+	consumedBytes = marshalUtil.ReadOffset()
+
+	return
+}
+
+// InputsFromMarshalUtil unmarshals a collection of Inputs using a MarshalUtil (for easier unmarshaling).
+func InputsFromMarshalUtil(marshalUtil *marshalutil.MarshalUtil) (inputs Inputs, err error) {
+	inputsCount, err := marshalUtil.ReadUint16()
+	if err != nil {
+		err = xerrors.Errorf("failed to parse inputs count (%v): %w", err, ErrParseBytesFailed)
+		return
+	}
+
+	var previousInput *Input
```

Comment:
Is the pointer here really needed?

Reply:
yes, because otherwise we are not able to determine the "unset" value. Of course you could introduce another bool to track that but that would not just be ugly but also slower :P

Reply:
I think it should work without the pointer in this case since an interface that does not hold a concrete value is `nil`. We do assign with `previousInput = &parsedInputs[i]` (effectively making `previousInput != nil` always true) but that happens after the initial check and then there should never be a `nil` concrete value be returned from `InputFromMarshalUtil`. See this example: https://play.golang.org/p/CzbZbUhLtXl

Reply:
Oh yeah you are right its an Interface! O_O